### PR TITLE
Add enum{a,b,c} field type for closed-set columns (issue #1030)

### PR DESCRIPTION
**Pull request by madmax983**
## Summary
Implements support for `enum{a,b,c}` field type across the code generation system, enabling scaffolding of closed-set columns end-to-end with database-level enforcement and type-safe Rust enums.

### DSL & Parsing
- Added `FieldKind::Enum` variant and `Field::variants` to represent closed-set fields
- Implemented `parse_enum_type()` to parse `enum{a,b,c}` syntax with validation
- Added `Field::enum_type_name()` and `Field::is_enum()` helper methods
- Updated `Field::rust_type()` to return the generated enum type name instead of `String` fallback

### Model Generation
- Implemented `render_enum_decl()` to generate Rust enum types with:
  - `Display`/`FromStr` traits for form round-tripping
  - `diesel::serialize::ToSql`/`deserialize::FromSql` impls for database storage as `TEXT`
  - `Default` derive (required by `UpdateX` patch struct)
  - `serde` support with `#[serde(rename)]` for variant mapping
- Added `validate_enum_field_collisions()` to prevent name collisions with model companion types and other enum fields
- Added `unquote_default_value()` helper for parsing `--default` values

### Schema & Migrations
- Added `enum_check_suffix()` to generate `CHECK (col IN ('a', 'b', ...))` constraints
- Updated `create_table_sql_with_metadata_and_id()`, `add_columns_up_sql()`, and `remove_columns_down_sql()` to emit CHECK constraints
- Constraints are preserved on migration rollback to maintain closed-set enforcement

### Scaffold Generation
- Added enum field validation in `parse_query_specs()` to reject `--query` on enum fields (not yet supported)
- Implemented `render_decoded_form()` enum handling with `FromStr` parsing and field-specific error messages
- Added enum field imports to routes file via `enum_import_suffix`
- Implemented `render_create_form_inputs()` and `render_edit_form_inputs()` to render `<select>` widgets with:
  - One `<option>` per variant
  - Humanized labels via new `humanize_label()` function
  - Proper `selected` binding in edit forms
- Added `render_enum_rejection_smoke_test()` to verify closed-set enforcement at both database and request-boundary layers
- Refactored SQL escaping into `escape_sql_for_rust_literal()` for reuse in test generation
- Enhanced reference stub table seeding to prevent FK constraint masking in tests

### Admin Generator
- Updated `admin_field_kind()` to recognize enum fields
- Modified `render_fields_vec()` to emit `AdminFieldKind::Select` for enum fields with variant options
- Extracted label humanization into shared `humanize_label()` function

### Validation & Error Handling
- Added enum field rejection in `generate job` (not supported in job args structs)
- Comprehensive error messages for invalid defaults, collisions, and unsupported operations
- Updated help text and `SUPPORTED_TYPES` constant to document `enum{a,b,...}` syntax

## Notable Implementation Details

- Enum variants are stored as `snake_case` in the database and Rust code, with `PascalCase` conversion for enum variant names
- Closed-set enforcement occurs at two layers: database `CHECK` constraint and Rust `FromStr` parsing in form handlers
- Smoke tests include both raw SQL insertion tests (database layer) and form submission tests (request boundary)
- The generated enum type name is derived from the field name via `pascal()`, enabling collision detection
- All enum fields require at least one variant; empty enums are rejected during parsing

https://claude.ai/code/session_01NtrprtZd928oqS82aLZpZz

### CODE REVIEW

File: autumn-cli/src/generate/scaffold.rs [hunk 1]
```diff
@@ -2242,13 +2385,166 @@ fn render_smoke_test(
...
 
-    if api {
+    let base = if api {
         render_api_smoke_test(plural, id_schema_type, &setup_calls)
     } else {
         render_index_smoke_test(pascal_name, plural, id_schema_type, &setup_calls)
+    };
+
+    if fields.iter().any(Field::is_enum) {
+        base + &render_enum_rejection_smoke_test(plural, fields, defaults, &setup_calls, api)
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/generate/scaffold.rs:2451]: Confirmed and fixed in bd2cc20 — the enum-rejection test's own copy of the setup now uses `CREATE TABLE IF NOT EXISTS <table>` so it doesn't race with the base index/api test's non-idempotent `CREATE TABLE` in the same shared `TestDb`. Added a regression test asserting the base test keeps the plain form (matching the real migration) while the enum test uses the idempotent form.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: autumn-cli/src/generate/model.rs [hunk 2]
```diff
@@ -743,6 +745,72 @@ fn validate_field_names(fields: &[Field]) -> Result<(), GenerateError> {
...
+fn validate_enum_field_collisions(
+    pascal_name: &str,
+    fields: &[Field],
+) -> Result<(), GenerateError> {
+    let mut seen_enum_types: Vec<(String, String)> = Vec::new();
+    for f in fields {
+        let Some(enum_ty) = f.enum_type_name() else {
+            continue;
+        };
+        let reserved = [
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/generate/model.rs:801]: Confirmed (and worse than described — it also breaks `.parse()`/`FromSql`/`String::new()` call sites, not just the type-position uses) and fixed in bd2cc20. Went with rejecting the reserved names (`String`, `Vec`, `Option`, `Result`, `Box`) rather than qualifying the template's std types, since the shadowing also breaks *other*, unrelated code in the same file once the enum's `use` is imported into the scaffold's routes file (e.g. `String::new()` in `decode_form`) — qualifying just this template's own paths wouldn't have covered that. Verified by generating `string:enum{a,b}` before/after the fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: autumn-cli/src/generate/scaffold.rs [hunk 3]
```diff
@@ -1297,7 +1357,7 @@ use autumn_web::ui::pagination::{{PagerOptions, pagination_nav}};
 use diesel::prelude::*;
 use diesel_async::RunQueryDsl;
 
-use crate::models::{snake_name}::{{{pascal_name}, New{pascal_name}, Update{pascal_name}}};
+use crate::models::{snake_name}::{{{pascal_name}, New{pascal_name}, Update{pascal_name}{enum_import_suffix}}};
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/generate/scaffold.rs:1416]: Confirmed and fixed in bd2cc20 — added a scaffold-only reserved-name check (`Path`, `Page`, `PageRequest`, `Bytes`, `CsrfFormField`, `CsrfToken`, `PagerOptions`, `Flash`, `ShardedDb`, `Db`, `AutumnError`, `AutumnResult`, `Markup`, `RunQueryDsl`) gated on `!--api` (routes.rs isn't emitted for API-only scaffolds, so the collision can't occur there). Verified by generating `path:enum{a,b}` before/after the fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: autumn-cli/src/generate/model.rs [hunk 4]
```diff
@@ -743,6 +745,95 @@ fn validate_field_names(fields: &[Field]) -> Result<(), GenerateError> {
...
+/// on being their real, unqualified meaning — `String::new()`, `s.parse()`
+/// returning a real `std::string::String`, `Vec<u8>` Bytea handling, etc. A
+/// field that pascalizes to one of these (e.g. `string:enum{a,b}` ->
+/// `pub enum String`) shadows the prelude import for the *entire file* it's
+/// declared in (and, once imported via `use crate::models::…::{…, String}`,
+/// for the entire scaffold routes file too), breaking unrelated code far
+/// from this DSL token — confirmed by generating `string:enum{a,b}` and
+/// observing `cargo check` fail with E0308/E0599 throughout the model and
+/// routes files.
+const PRELUDE_TYPE_NAMES: &[&str] = &["String", "Vec", "Option", "Result", "Box"];
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/generate/model.rs]: Confirmed and fixed in d8fabef — added `Into` to the reserved prelude-type list (`String`, `Vec`, `Option`, `Result`, `Box`, `Into`) checked in `validate_enum_field_collisions`, same guard that already covered the type-position shadows. Verified by generating `into:enum{a,b}` before the fix (fails with the described `Into::into` resolution error) and after (rejected at generate time with a pointer back to the field).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: autumn-cli/src/generate/admin.rs [hunk 5]
```diff
@@ -629,13 +624,29 @@ fn render_fields_vec(
...
+        let effective_select: Option<SelectSpec> = options
+            .select
+            .iter()
+            .find(|s| s.field == f.name)
+            .cloned()
+            .or_else(|| {
+                f.is_enum().then(|| SelectSpec {
+                    field: f.name.clone(),
+                    values: f.variants.clone(),
+                })
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/generate/admin.rs:642]: Confirmed and fixed in d8fabef — the smoke-test form-body builder now uses the field's own first declared variant (`f.variants.first()`) as the sample value for enum fields instead of the literal `"test"`, falling back to `"test"` for non-enum fields. Added a regression test asserting `status=draft` appears and `status=test` does not for an enum-backed admin resource.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: autumn-cli/src/generate/scaffold.rs [hunk 6]
```diff
@@ -350,6 +353,58 @@ pub fn run(name: &str, field_tokens: &[String], flags: Flags, options: &Scaffold
...
+    "CsrfFormField",
+    "CsrfToken",
+    "PagerOptions",
+    "Flash",
+    "ShardedDb",
+    "Db",
+    "AutumnError",
+    "AutumnResult",
+    "Markup",
+    "RunQueryDsl",
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/generate/scaffold.rs:383]: Confirmed and fixed in 0e5f3a1 — added `ToString` to `ROUTES_FILE_RESERVED_NAMES`, checked by `validate_enum_field_names_against_routes_imports` (same guard that already covered `Path`/`Page`/etc.). Reproduced the exact bug first with a red test (`to_string:Option<enum{a,b}>` generating `pub enum ToString` and the unqualified `ToString::to_string` path in routes.rs), then made it pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: autumn-cli/src/generate/scaffold.rs [hunk 7]
```diff
@@ -2242,13 +2441,183 @@ fn render_smoke_test(
...
+    // "relation already exists" — make *this* test's own copy of the
+    // `CREATE TABLE` idempotent so it works whether it runs alongside the
+    // base test or standalone (e.g. `cargo test posts_rejects_out_of_set_status
+    // -- --ignored`). The reference-stub-table `CREATE TABLE`s are already
+    // `IF NOT EXISTS` (see `render_reference_stub_tables_sql`), so only the
+    // main table's statement needs the same treatment.
+    let setup_calls = setup_calls.replacen(
+        &format!("CREATE TABLE {plural} ("),
+        &format!("CREATE TABLE IF NOT EXISTS {plural} ("),
+        1,
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/generate/scaffold.rs:2557]: Confirmed and fixed in 4759cd7 — extended the same rewrite to replace every `CREATE INDEX idx_` with `CREATE INDEX IF NOT EXISTS idx_` in the enum-rejection test's own copy of `setup_calls` (the `CREATE TABLE` rewrite is unchanged). Reproduced first with a red test (indexed `title` field + enum `status` field scaffold, asserting the enum test's setup is idempotent while the base test's stays a plain `CREATE INDEX`), then made it pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: autumn-cli/src/generate/dsl.rs [hunk 8]
```diff
@@ -361,9 +415,81 @@ pub fn parse_field(token: &str) -> Result<Field, GenerateError> {
...
+        }
+        if !is_valid_ident(variant) {
+            return Err(format!("'{variant}' is not a valid snake_case identifier"));
+        }
+        if is_rust_keyword(variant) {
+            return Err(format!(
+                "'{variant}' is a Rust keyword and cannot be used as an enum variant"
+            ));
+        }
+        let pascal = naming::pascal(variant);
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/generate/dsl.rs:475]: Confirmed and fixed in 4759cd7 — `parse_enum_type` now rejects a variant whose `pascal()` result is empty or starts with an ASCII digit, in addition to the existing `is_valid_ident`/keyword checks. Reproduced first with red tests for both failure modes (`_2fa` → `2fa`, leading digit; `__` → empty string), then made them pass.

---
_Generated by [Claude Code](https://claude.ai/code)_
